# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/websocket/WebSocketConnectionParent.h
# Commit: f7b6aac1b58b
# Full Hash: f7b6aac1b58bc959e431136e1b23231abe55d0b1
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2020-07-16 21:27:37
# Regressor Bug: 1497249
# File Overlap Count: 4
# Description:
#   Bug 1497249 - P2: ipdl for nsIWebSocketConnection r=michal
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D30624
# ==============================================================================

diff -r 74bb3c2ab767 -r f7b6aac1b58b netwerk/protocol/websocket/WebSocketConnectionParent.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/netwerk/protocol/websocket/WebSocketConnectionParent.h	Thu Jul 16 13:34:58 2020 +0000
@@ -0,0 +1,67 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set sw=2 ts=8 et ft=cpp : */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_net_WebSocketConnectionParent_h
+#define mozilla_net_WebSocketConnectionParent_h
+
+#include "mozilla/net/PWebSocketConnectionParent.h"
+#include "nsISupportsImpl.h"
+#include "nsIWebSocketConnection.h"
+
+class nsIHttpUpgradeListener;
+
+namespace mozilla {
+namespace net {
+
+// WebSocketConnectionParent implements nsIWebSocketConnection and provides
+// interface for WebSocketChannel to send/receive data. The ownership model for
+// TransportProvider is that IPDL and WebSocketChannel hold strong reference of
+// WebSocketConnectionParent. When nsIWebSocketConnection::Close is called, a
+// __delete__ message will be sent and the IPC actor will be deallocated as
+// well.
+
+#define WEB_SOCKET_CONNECTION_PARENT_IID             \
+  {                                                  \
+    0x1cc3cb61, 0x0c09, 0x4f58, {                    \
+      0x9a, 0x64, 0x44, 0xf7, 0x92, 0x86, 0xbc, 0x00 \
+    }                                                \
+  }
+
+class WebSocketConnectionParent final : public PWebSocketConnectionParent,
+                                        public nsIWebSocketConnection {
+ public:
+  NS_DECL_THREADSAFE_ISUPPORTS
+  NS_DECL_NSIWEBSOCKETCONNECTION
+  NS_DECLARE_STATIC_IID_ACCESSOR(WEB_SOCKET_CONNECTION_PARENT_IID)
+
+  explicit WebSocketConnectionParent(nsIHttpUpgradeListener* aListener);
+
+  mozilla::ipc::IPCResult RecvOnTransportAvailable(
+      const nsCString& aSecurityInfoSerialization);
+  mozilla::ipc::IPCResult RecvOnError(const nsresult& aStatus);
+  mozilla::ipc::IPCResult RecvOnTCPClosed();
+  mozilla::ipc::IPCResult RecvOnDataReceived(nsTArray<uint8_t>&& aData);
+  mozilla::ipc::IPCResult RecvOnUpgradeFailed(const nsresult& aReason);
+
+  void ActorDestroy(ActorDestroyReason aWhy) override;
+
+ private:
+  virtual ~WebSocketConnectionParent();
+
+  nsCOMPtr<nsIHttpUpgradeListener> mUpgradeListener;
+  nsCOMPtr<nsIWebSocketConnectionListener> mListener;
+  nsCOMPtr<nsIEventTarget> mEventTarget;
+  nsCOMPtr<nsISupports> mSecurityInfo;
+  bool mClosed;
+};
+
+NS_DEFINE_STATIC_IID_ACCESSOR(WebSocketConnectionParent,
+                              WEB_SOCKET_CONNECTION_PARENT_IID)
+
+}  // namespace net
+}  // namespace mozilla
+
+#endif  // mozilla_net_WebSocketConnectionParent_h